doc/man/man3/ProcessImages.3
.TH "ProcessImages" 3 "Wed Jan 19 2022" "Version v1.0" "CV" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ProcessImages
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <ProcessImages\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBinitLabels\fP ()"
.br
.RI "reads in dataset_labeling\&.txt and sets total amount of license plates and characters that can be recognized\&. "
.ti -1c
.RI "void \fBpreprocessImage\fP (cv::Mat \fBimage\fP)"
.br
.RI "Function to apply multiple filters to the original picture, so that it is ready for extracting all of its contours\&. The results are saved in the matrixes of this class\&. (Note that there are some that are not being used, we tried out several methods to preprocess our pictures) "
.ti -1c
.RI "void \fBexecute\fP ()"
.br
.RI "Function to start the license plate recognition, it loops over all existing pictures from the dataset and calls necessary functions of this class\&. Also adds height and width to labels\&. "
.ti -1c
.RI "void \fBprintFilterConditions\fP (int i)"
.br
.RI "prints filter conditions for the current picture that is analyzed e\&.g Min Area Outer Contour (Rectangle)\&. "
.ti -1c
.RI "void \fBfilterImageData\fP ()"
.br
.RI "Set width, height of a picture and calculate relative min and max Area for outer contours in relation with image width and height\&. "
.ti -1c
.RI "void \fBfilterContours\fP (bool brighter)"
.br
.RI "First we start by extracting all of the contours of a picture, secondly the perimeter of a contour is calculated and the contour is approxymated (for instance a rectangle can be drown with 4 vertices)\&. After that we check if the contour in fact contains only 4 points, indicating that it is a rectangle\&. We filter the extracted rectangle by area and apect ratio\&. The next step is to see how many white pixels exist in the contour, since a license plate will consist of a lot of white pixels and some black pixels indicating the letters\&. If the amount of white pixels is high enough we again filter for all countours that exist in our rectangle\&. In this step we cant to see if we cann figure out the letters on our license plate\&. If the contour has a specific area we let it pass and count it\&. Since a license plate will have about 15 letters on it we check if we found 4-12 contours that fit our area check\&. If so we sucessfully found a licenseplate on the picture\&. "
.ti -1c
.RI "void \fBsetFilterConditionsInnerContours\fP (cv::Mat &subImgCopy)"
.br
.RI "Calculate relative min and max area for inner Contour\&. "
.ti -1c
.RI "void \fBdisplayResults\fP (cv::Mat &image_copy)"
.br
.RI "Display the picture with recognized license plates and the corresponding threshold image\&. "
.ti -1c
.RI "void \fBcountInnerContours\fP (unsigned int &contourCounter, int i, cv::Mat &subImgCopy)"
.br
.RI "When outer the outer contour survives all of our check, count all contours inside the rectangle with a specific area\&. "
.ti -1c
.RI "void \fBprintCandidateData\fP (int i, cv::Rect2f &resultRect, double whitePixelsInRectangle, float proportionAreaWhitepixels)"
.br
.RI "print Area, Number of white pixels and ratio of white and black pixels in rectangle "
.ti -1c
.RI "void \fBisRectangle\fP (cv::Rect2f &resultRect, int &retflag)"
.br
.RI "Check aspect ratio and area of cadidate\&. Ignore if not in these bounderies\&. "
.ti -1c
.RI "bool \fBisImageBright\fP (cv::Mat \fBimage\fP)"
.br
.RI "Change from BGR to HSV and check value across all pixels\&. Decide from value if image is bright or not\&. Also save brightness in labels\&. "
.ti -1c
.RI "void \fBmakeBrighter\fP (double alpha, int beta)"
.br
.RI "Make image brighter with alpha and beta values\&. "
.ti -1c
.RI "void \fBgammaCorrection\fP (const cv::Mat &img, const double gamma_)"
.br
.RI "Make image brighter with gamma value\&. "
.ti -1c
.RI "void \fBgetDesktopResolution\fP ()"
.br
.RI "store desktop resolution "
.ti -1c
.RI "void \fBprocess\fP ()"
.br
.ti -1c
.RI "float \fBdistanceBetweenTwoPoints\fP (float x1, float y1, float x2, float y2)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "std::string \fBimageToRead\fP"
.br
.ti -1c
.RI "int \fBddepth\fP = CV_8U"
.br
.ti -1c
.RI "int \fBdesktopWidth\fP"
.br
.ti -1c
.RI "int \fBdesktopHeigth\fP"
.br
.ti -1c
.RI "\fBMetric\fP \fBmetric\fP"
.br
.ti -1c
.RI "int \fBindex\fP = 0"
.br
.ti -1c
.RI "float \fBminAreaOuter\fP"
.br
.ti -1c
.RI "float \fBmaxAreaOuter\fP"
.br
.ti -1c
.RI "float \fBminAreaInner\fP"
.br
.ti -1c
.RI "float \fBmaxAreaInner\fP"
.br
.ti -1c
.RI "bool \fBDEBUG_FLAG_SHOW_RESULT\fP = false"
.br
.ti -1c
.RI "cv::Mat \fBimage\fP"
.br
.ti -1c
.RI "cv::Mat \fBimageGrey\fP"
.br
.ti -1c
.RI "cv::Mat \fBimageBlurred\fP"
.br
.ti -1c
.RI "cv::Mat \fBimageThresh\fP"
.br
.ti -1c
.RI "cv::Mat \fBimageThreshResized\fP"
.br
.ti -1c
.RI "cv::Mat \fBsobelX\fP"
.br
.ti -1c
.RI "cv::Mat \fBsobelY\fP"
.br
.ti -1c
.RI "cv::Mat \fBsobelResult\fP"
.br
.ti -1c
.RI "cv::Mat \fBerodeResult\fP"
.br
.ti -1c
.RI "cv::Mat \fBdilateResult\fP"
.br
.ti -1c
.RI "cv::Mat \fBimageConcat\fP"
.br
.ti -1c
.RI "cv::Mat \fBresizedDown\fP"
.br
.ti -1c
.RI "cv::Mat \fBres\fP"
.br
.ti -1c
.RI "cv::Mat \fBnewImage\fP"
.br
.ti -1c
.RI "std::vector< std::vector< cv::Point > > \fBcontours\fP"
.br
.ti -1c
.RI "std::vector< std::vector< cv::Point > > \fBcontoursSecond\fP"
.br
.ti -1c
.RI "std::vector< cv::Vec4i > \fBhierarchy\fP"
.br
.ti -1c
.RI "std::vector< cv::Vec4i > \fBhierarchySecond\fP"
.br
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "void ProcessImages::countInnerContours (unsigned int & contourCounter, int i, cv::Mat & subImgCopy)"

.PP
When outer the outer contour survives all of our check, count all contours inside the rectangle with a specific area\&. 
.PP
\fBParameters\fP
.RS 4
\fIcontourCounter\fP 
.br
\fIi\fP 
.br
\fIsubImgCopy\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::displayResults (cv::Mat & image_copy)"

.PP
Display the picture with recognized license plates and the corresponding threshold image\&. 
.PP
\fBParameters\fP
.RS 4
\fIimage_copy\fP 
.br
 
.RE
.PP

.SS "float ProcessImages::distanceBetweenTwoPoints (float x1, float y1, float x2, float y2)"

.SS "void ProcessImages::execute ()"

.PP
Function to start the license plate recognition, it loops over all existing pictures from the dataset and calls necessary functions of this class\&. Also adds height and width to labels\&. 
.SS "void ProcessImages::filterContours (bool brighter)"

.PP
First we start by extracting all of the contours of a picture, secondly the perimeter of a contour is calculated and the contour is approxymated (for instance a rectangle can be drown with 4 vertices)\&. After that we check if the contour in fact contains only 4 points, indicating that it is a rectangle\&. We filter the extracted rectangle by area and apect ratio\&. The next step is to see how many white pixels exist in the contour, since a license plate will consist of a lot of white pixels and some black pixels indicating the letters\&. If the amount of white pixels is high enough we again filter for all countours that exist in our rectangle\&. In this step we cant to see if we cann figure out the letters on our license plate\&. If the contour has a specific area we let it pass and count it\&. Since a license plate will have about 15 letters on it we check if we found 4-12 contours that fit our area check\&. If so we sucessfully found a licenseplate on the picture\&. 
.PP
\fBParameters\fP
.RS 4
\fIbrighter\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::filterImageData ()"

.PP
Set width, height of a picture and calculate relative min and max Area for outer contours in relation with image width and height\&. 
.SS "void ProcessImages::gammaCorrection (const cv::Mat & img, const double gamma_)"

.PP
Make image brighter with gamma value\&. 
.PP
\fBParameters\fP
.RS 4
\fIimg\fP 
.br
\fIgamma_\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::getDesktopResolution ()"

.PP
store desktop resolution 
.SS "void ProcessImages::initLabels ()"

.PP
reads in dataset_labeling\&.txt and sets total amount of license plates and characters that can be recognized\&. 
.SS "bool ProcessImages::isImageBright (cv::Mat image)"

.PP
Change from BGR to HSV and check value across all pixels\&. Decide from value if image is bright or not\&. Also save brightness in labels\&. 
.PP
\fBParameters\fP
.RS 4
\fIimage\fP 
.RE
.PP
\fBReturns\fP
.RS 4

.br
 
.RE
.PP

.SS "void ProcessImages::isRectangle (cv::Rect2f & resultRect, int & retflag)"

.PP
Check aspect ratio and area of cadidate\&. Ignore if not in these bounderies\&. 
.PP
\fBParameters\fP
.RS 4
\fIresultRect\fP 
.br
\fIretflag\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::makeBrighter (double alpha, int beta)"

.PP
Make image brighter with alpha and beta values\&. 
.PP
\fBParameters\fP
.RS 4
\fIalpha\fP 
.br
\fIbeta\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::preprocessImage (cv::Mat image)"

.PP
Function to apply multiple filters to the original picture, so that it is ready for extracting all of its contours\&. The results are saved in the matrixes of this class\&. (Note that there are some that are not being used, we tried out several methods to preprocess our pictures) 
.PP
\fBParameters\fP
.RS 4
\fIimage\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::printCandidateData (int i, cv::Rect2f & resultRect, double whitePixelsInRectangle, float proportionAreaWhitepixels)"

.PP
print Area, Number of white pixels and ratio of white and black pixels in rectangle 
.PP
\fBParameters\fP
.RS 4
\fIi\fP 
.br
\fIresultRect\fP 
.br
\fIwhitePixelsInRectangle\fP 
.br
\fIproportionAreaWhitepixels\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::printFilterConditions (int i)"

.PP
prints filter conditions for the current picture that is analyzed e\&.g Min Area Outer Contour (Rectangle)\&. 
.PP
\fBParameters\fP
.RS 4
\fIi\fP 
.br
 
.RE
.PP

.SS "void ProcessImages::process ()"

.SS "void ProcessImages::setFilterConditionsInnerContours (cv::Mat & subImgCopy)"

.PP
Calculate relative min and max area for inner Contour\&. 
.PP
\fBParameters\fP
.RS 4
\fIsubImgCopy\fP 
.br
 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "std::vector<std::vector<cv::Point> > ProcessImages::contours"

.SS "std::vector<std::vector<cv::Point> > ProcessImages::contoursSecond"

.SS "int ProcessImages::ddepth = CV_8U"

.SS "bool ProcessImages::DEBUG_FLAG_SHOW_RESULT = false"

.SS "int ProcessImages::desktopHeigth"

.SS "int ProcessImages::desktopWidth"

.SS "cv::Mat ProcessImages::dilateResult"

.SS "cv::Mat ProcessImages::erodeResult"

.SS "std::vector<cv::Vec4i> ProcessImages::hierarchy"

.SS "std::vector<cv::Vec4i> ProcessImages::hierarchySecond"

.SS "cv::Mat ProcessImages::image"

.SS "cv::Mat ProcessImages::imageBlurred"

.SS "cv::Mat ProcessImages::imageConcat"

.SS "cv::Mat ProcessImages::imageGrey"

.SS "cv::Mat ProcessImages::imageThresh"

.SS "cv::Mat ProcessImages::imageThreshResized"

.SS "std::string ProcessImages::imageToRead"

.SS "int ProcessImages::index = 0"

.SS "float ProcessImages::maxAreaInner"

.SS "float ProcessImages::maxAreaOuter"

.SS "\fBMetric\fP ProcessImages::metric"

.SS "float ProcessImages::minAreaInner"

.SS "float ProcessImages::minAreaOuter"

.SS "cv::Mat ProcessImages::newImage"

.SS "cv::Mat ProcessImages::res"

.SS "cv::Mat ProcessImages::resizedDown"

.SS "cv::Mat ProcessImages::sobelResult"

.SS "cv::Mat ProcessImages::sobelX"

.SS "cv::Mat ProcessImages::sobelY"


.SH "Author"
.PP 
Generated automatically by Doxygen for CV from the source code\&.


OpenCVTest/ProcessImages.cpp
#include "ProcessImages.h"

void ProcessImages::initLabels() {
    metric.readLabels();
    metric.setTotalAmountLicensePlatesOnPictures();
    metric.setTotalAmountOfCharactersOnPictures();
    metric.printLabels();
}

void ProcessImages::preprocessImage(cv::Mat image) {
    cv::erode(image, erodeResult, cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2, 2)));
    cv::dilate(erodeResult, dilateResult, cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2, 2)));

    //cv::dilate(image, dilateResult, cv::getStructuringElement(cv::MORPH_RECT, cv::Size(1, 1)));
    //cv::erode(dilateResult, erodeResult, cv::getStructuringElement(cv::MORPH_RECT, cv::Size(1, 1)));

    //cv::resize(dilateResult, resizedDown, cv::Size(down_width, down_height), cv::INTER_LINEAR);

    cv::cvtColor(dilateResult, imageGrey, cv::COLOR_BGR2GRAY);

    cv::GaussianBlur(imageGrey, imageBlurred, cv::Size(3, 3), 0);

    cv::threshold(imageBlurred, imageThresh, 129, 255, cv::THRESH_BINARY);

    //cv::Sobel(imageThresh, sobel_x, ddepth, 1, 0, 3, cv::BORDER_DEFAULT);
    //cv::Sobel(imageThresh, sobel_y, ddepth, 0, 1, 3, cv::BORDER_DEFAULT);
    //cv::magnitude(sobel_x, sobel_y, sobelResult);

    //cv::hconcat(image, imageThresh, imageConcat);
}

void ProcessImages::execute() {
    getDesktopResolution();
    process();
    metric.printRecognizedOutOfAll();
    cv::destroyAllWindows();
}

void ProcessImages::process() {
    for (int i = 1; i < NUMBER_OF_IMAGES + 1; i++) {
        index = i - 1;
        std::string imageToRead = "images/" + std::to_string(i) + ".png";

        image = cv::imread(imageToRead);

        if (image.empty()) {
            imageToRead = "images/" + std::to_string(i) + ".jpg";
            image = cv::imread(imageToRead);

            if (image.empty()) {
                std::cout << "Image not Found !!!" << std::endl;
                std::cin.get();
                continue;
            }
        }

        filterImageData();
        printFilterConditions(i);

        if (isImageBright(image)) {
            //std::cout << "Image is Bright" << std::endl;
            preprocessImage(image);
            filterContours(true);
        }
        else {
            //std::cout << "Image is not Bright ! " << std::endl;
            makeBrighter(2.0, 10);
            gammaCorrection(image, 0.5);
            preprocessImage(res);
            preprocessImage(newImage);
            filterContours(false);
        }
    }
}

void ProcessImages::printFilterConditions(int i)
{
    std::cout << "[Filter Conditions for Picture " << i << "]\n";
    std::cout << "Min Area outer Contour : " << minAreaOuter << " --- " << "Max Area outer Contour : " << maxAreaOuter << "\n";
    std::cout << "White Pixel Threshold lower boundary for rectangle : " << WHITE_PIXEL_TRESHOLD_LOW << "\n";
    std::cout << "White Pixel Threshold upper boundary for rectangle : " << WHITE_PIXEL_THRESHOLD_HIGH << "\n";
    std::cout << "Min aspect ratio for rect : " << ASPECT_RATIO_MIN << " --- " << "Max aspect ratio for rect : " << ASPECT_RATIO_MAX << std::endl;
    std::cout << "Darkness Threshold for whole picture : " << DARKNESS_THRESHOLD << "\n\n";
}

void ProcessImages::filterImageData()
{
    metric.labels.at(index).width = image.rows;
    metric.labels.at(index).heigth = image.cols;

    minAreaOuter = (image.rows / 50.0f) * (image.cols / 50.0f);
    maxAreaOuter = (image.rows / 8.0f) * (image.cols / 8.0f);
}

void ProcessImages::filterContours(bool brighter) {
    float proportionAreaWhitepixels = 0.f;
    cv::Mat image_copy;
    std::vector<cv::Point> approx;
    double perimeter = 0.0;
  
    if (brighter) {
        image_copy = image.clone();
    }
    else {
        image_copy = newImage.clone();
    }
    
    //Draw optimal
    for (int h = 0; h < metric.labels.at(index).coordinates.size(); h++) {
        cv::rectangle(image_copy, cv::Rect(metric.labels.at(index).coordinates.at(h).x, metric.labels.at(index).coordinates.at(h).y, 50, 50), cv::Scalar(0, 0, 255), 2);
    }

    cv::findContours(imageThresh, contours, hierarchy, cv::RETR_TREE, cv::CHAIN_APPROX_NONE); // What influence do RETR_TREE and CHAIN_APPROX_NONE have on the algorithm
    
    metric.recognizedPerPicture = 0;
    metric.falsePositiveCounterPerPicture = 0;
    std::vector<int> arrayRecognizedCharacters;

    for (int i = 0; i < contours.size(); i++) {
        perimeter = cv::arcLength(contours.at(i), true);
        cv::approxPolyDP(contours.at(i), approx, 0.04 * perimeter, true);

        if (approx.size() == 4) {
            cv::Rect2f resultRect = cv::boundingRect(approx);

            if (!isRectangle(resultRect)) {
                //std::cout << "Eliminated not recognized as rectangle ! Area or Ratio does not fit" << std::endl;
                continue;
            }

            cv::Mat subImg = imageThresh(resultRect);
            double whitePixelsInRectangle = cv::countNonZero(subImg);
            proportionAreaWhitepixels = (whitePixelsInRectangle / (resultRect.area()));
            printCandidateData(i, resultRect, whitePixelsInRectangle, proportionAreaWhitepixels);
            
            if (proportionAreaWhitepixels < WHITE_PIXEL_TRESHOLD_LOW || proportionAreaWhitepixels > WHITE_PIXEL_THRESHOLD_HIGH) {
                std::cout << "Eliminated to high or to low white pixel ratio !" << "\n\n";
                continue;
            }
            
            cv::findContours(subImg, contoursSecond, hierarchySecond, cv::RETR_TREE, cv::CHAIN_APPROX_NONE);
            cv::Mat subImgCopy = image_copy(resultRect);
            cv::drawContours(subImgCopy, contoursSecond, -1, cv::Scalar(0, 255, 255), 1, cv::LINE_AA);

            setFilterConditionsInnerContours(subImgCopy);

            unsigned int contourCounter = 0;
            countInnerContours(contourCounter, i, subImgCopy);

            if (contourCounter < 4 || contourCounter > 12) {
                std::cout << "Eliminated to many or to little contours counted !" << "\n\n";
                continue;
            }
            std::cout << "\n[Candidate " << std::to_string(i) + "]" << " Characters on license plate: " << contourCounter - 1 << "\n\n";
            cv::rectangle(image_copy, resultRect, cv::Scalar(0, 255, 0), 2);
            
            bool isFalsePositive = false;

            for (int h = 0; h < metric.labels.at(index).coordinates.size(); h++) {
                float d = distanceBetweenTwoPoints(resultRect.x, resultRect.y, metric.labels.at(index).coordinates.at(h).x, metric.labels.at(index).coordinates.at(h).y);
                std::cout << "Distance : " << d << std::endl;

                if (d > FALSE_POSITIVE_THRESHOLD) {
                    isFalsePositive = true;
                }
                if (d < FALSE_POSITIVE_THRESHOLD) {
                    isFalsePositive = false;
                    break;
                }
            }
            
            if (isFalsePositive) {
                metric.falsePositiveCounterPerPicture++;
                contourCounter = 0;
                break;
            }

            arrayRecognizedCharacters.push_back(contourCounter);
            metric.recognizedCharacters += contourCounter;
            metric.recognizedLicensePlates++;
            metric.recognizedPerPicture++;

            if (metric.recognizedPerPicture > metric.labels.at(index).numberplateCount) {
                metric.recognizedLicensePlates -= metric.recognizedPerPicture;
                metric.recognizedCharacters -= contourCounter;
                break;
            }
        }
    }
    metric.recognizedCharactersPerLicenseplate.push_back(arrayRecognizedCharacters);
    metric.printMetricOnePicture(index);
    if (DEBUG_FLAG_SHOW_RESULT) {
        displayResults(image_copy);
        cv::destroyAllWindows();
    }
}

void ProcessImages::setFilterConditionsInnerContours(cv::Mat& subImgCopy)
{
    minAreaInner = (subImgCopy.rows / 6.0f) * (subImgCopy.cols / 6.0f);
    maxAreaInner = (subImgCopy.rows / 3.2f) * (subImgCopy.cols / 3.2f);
    std::cout << "Min Area inner Contour : " << minAreaInner << " --- " << "Max Area inner Contour : " << maxAreaInner << "\n";
}

void ProcessImages::displayResults(cv::Mat& image_copy)
{
    float resizeFactor = 0.6;
    if (image_copy.rows > 3000.0f || image_copy.cols > 3000.0f) {
        resizeFactor = 0.25;
    }
    cv::resize(image_copy, resizedDown, cv::Size(image_copy.rows * resizeFactor, image_copy.cols * resizeFactor), cv::INTER_LINEAR);
    cv::resize(imageThresh, imageThreshResized, cv::Size(imageThresh.rows * resizeFactor, imageThresh.cols * resizeFactor), cv::INTER_LINEAR);
    cv::namedWindow("Result");
    cv::namedWindow("Thresh");
    cv::moveWindow("Result", 0, 0);
    cv::moveWindow("Thresh", desktopWidth / 2, 0);
    cv::imshow("Result", resizedDown);
    cv::imshow("Thresh", imageThreshResized);
    cv::waitKey(0);
}

void ProcessImages::countInnerContours(unsigned int& contourCounter, int i, cv::Mat& subImgCopy)
{
    for (int j = 0; j < contoursSecond.size(); j++) {
        double perimeter = cv::arcLength(contoursSecond.at(j), true);
        std::vector<cv::Point> approx;
        cv::approxPolyDP(contoursSecond.at(j), approx, 0.04 * perimeter, true);
        cv::Rect2f resultRect = cv::boundingRect(approx);

        if(resultRect.area() < minAreaInner || resultRect.area() > maxAreaInner) {
            //std::cout << "Eliminated area of inner Contour does not match min or max !" << std::endl;
            continue;
        }
        contourCounter++;
        std::cout << "[Candidate " + std::to_string(i) + "]" << " Area of Contour : " << resultRect.area() << " --- " << "Perimeter : " << perimeter << "ContourCounter : " << contourCounter << std::endl;
        cv::Mat character = subImgCopy(resultRect);
    }
}

void ProcessImages::printCandidateData(int i, cv::Rect2f& resultRect, double whitePixelsInRectangle, float proportionAreaWhitepixels)
{
    std::cout << "[Candidate " + std::to_string(i) + "]" << " Area: " << resultRect.area() << std::endl;
    std::cout << "[Candidate " + std::to_string(i) + "]" << " Number of white pixels: " << whitePixelsInRectangle << std::endl;
    std::cout << "[Candidate " + std::to_string(i) + "]" << " Ratio white pixel to area of rectangle : " << proportionAreaWhitepixels << "\n";

}

bool ProcessImages::isRectangle(cv::Rect2f& resultRect)
{
    float aspectratio = resultRect.width / resultRect.height;
    if (aspectratio < 1.0f) {
        aspectratio = 1 / aspectratio;
    }

    if ((aspectratio < ASPECT_RATIO_MIN || aspectratio > ASPECT_RATIO_MAX) || (resultRect.area() < minAreaOuter || resultRect.area() > maxAreaOuter)){
        return false;
    }
    return true;
}

bool ProcessImages::isImageBright(cv::Mat image) {
    cv::Mat hsv;
    cv::Mat copy = image.clone();
    cv::cvtColor(copy, hsv, cv::COLOR_BGR2HSV);

    const auto result = cv::mean(hsv);

    metric.labels.at(index).brightness = result[2];

    if (result[2] < DARKNESS_THRESHOLD) {
        return 0;
    }
    else {
        return 1;
    }

}

void ProcessImages::makeBrighter(double alpha, int beta) {
    newImage = cv::Mat::zeros(image.size(), image.type());

    for (int y = 0; y < image.rows; y++) {
        for (int x = 0; x < image.cols; x++) {
            for (int c = 0; c < image.channels(); c++) {
                newImage.at<cv::Vec3b>(y, x)[c] = cv::saturate_cast<uchar>(alpha * image.at<cv::Vec3b>(y, x)[c] + beta);
            }
        }
    }

    //cv::imshow("Original Image", image);
    //cv::imshow("New Image", new_image);

    //cv::waitKey();
}

void ProcessImages::gammaCorrection(const cv::Mat& img, const double gamma_) {
    CV_Assert(gamma_ >= 0);

    cv::Mat lookUpTable(1, 256, CV_8U);
    uchar* p = lookUpTable.ptr();

    for (int i = 0; i < 256; ++i) {
        p[i] = cv::saturate_cast<uchar>(cv::pow(i / 255.0, gamma_) * 255.0);
    }

    res = img.clone();
    cv::LUT(img, lookUpTable, res);
    // cv::imshow("Gamma correction", res);
}

void ProcessImages::getDesktopResolution() {
    RECT desktop;
    const HWND hDesktop = GetDesktopWindow();
    GetWindowRect(hDesktop, &desktop);
    desktopWidth = desktop.right;
    desktopHeigth = desktop.bottom;
}

float ProcessImages::distanceBetweenTwoPoints(float x1, float y1, float x2, float y2){
    float diffX = x2 - x1;
    float diffY = y2 - y1;

    return sqrtf(pow(diffX, 2.0f) + pow(diffY, 2.0f));
}